src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentCopy = {};
+    } else if (action.type === 'addProperties') {
+      currentCopy = { ...currentCopy, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentCopy = { ...currentCopy };
+
+      for (const key of action.keysToRemove) {
+        delete currentCopy[key];
+      }
+    }
```

Comment:
Instead of using multiple `if...else if` statements, consider using a `switch` statement for better readability and maintainability, especially since you have a limited amount of conditions based on the `action.type`.